pyproject.toml

```diff
@@ -5,3 +5,29 @@ build-backend = "setuptools.build_meta"
 [tool.black]
 line-length = 120
 target-version = ['py37', 'py38', 'py39', 'py310']
+
+[tool.mypy]
+exclude = ['docs/']
+allow_redefinition = false
+check_untyped_defs = true
+disallow_untyped_decorators = true
+disallow_any_explicit = false
+disallow_any_generics = true
+disallow_untyped_calls = true
+ignore_errors = false
+ignore_missing_imports = true
+implicit_reexport = false
+strict_optional = true
+strict_equality = true
+no_implicit_optional = true
+warn_unused_ignores = true
+warn_redundant_casts = true
+warn_unused_configs = true
+warn_unreachable = true
+warn_no_return = true
```

Comment:
Not a big fan of this one

Reply:
why? this is quite useful because it could help catch code paths with missing return statements

Reply:
I don't really see what problem solves. If there is no return statement, None is returned by default.

Reply:
It solves at least 2 problems:

1. We forgot to return something in one code path (e.g. in some `elif` block) by accident - that's a bug
2. According to the Zen of Python "explicit is better than implicit", so instead of relying on implicit `return None` let's write it explicitly. From my experience, usually implicit `return None` leads to confusion.

Reply:
1. Let's look at the `elif` example:

```python
def my_function(x: int) -> int:
    if x < 0:
        ...
        return -1
    elif x < 50:
        ...
    elif x < 100:
        return 100
    # here you could `raise SomeException()` or `return SomeSentinel()` or just do nothing
```

Then the code is valid and will execute but it doesn't work as expected because when `x < 50` it will return `None` (raise an exception or return some sentinel value), not an `int` instance like declared in type hinting (or generally not an expected value).

2. This is mostly about readability and beaing explicit:

```python
def my_function(x: int) -> int | None:
    if x > 0:
        return x * 10
    return None
```
is more readable and explicit (you don't have to remember that python always returns `None` implicitly, which sometimes could cost a lot of time when the function is complicated) than:

```python
def my_function(x: int) -> int | None:
    if x > 0:
        return x * 10
```

Reply:
I expected that mypy would return an error when checking this function using `warn_no_return = false`, simply because there is a branch (the one that fails all the checks) that returns None, since it's the implicit return.
```python
def my_function(x: int) -> int:
    if x < 0:
        return -1
    elif x < 50:
        return 50
    elif x < 100:
        return 100
```
But [it doesn't](https://mypy-play.net/?mypy=latest&python=3.10&gist=e7ec5c354ba293aa9f2a8f0ac7350005&flags=no-warn-no-return), and frankly I don't understand why. Instead [enabling `warn_no_return=true`](https://mypy-play.net/?mypy=latest&python=3.10&gist=17f048d45331a139515747021312531a&flags=no-warn-no-return), or [returning None explicitly](https://mypy-play.net/?mypy=latest&python=3.10&gist=96675cf9cdb7c505b6bcca5c1404ea5e&flags=no-warn-no-return), it correctly shows an error. I don't understand this behaviour at all, there is a branch of the function returning None (implicitly), and IMO mypy should show an error regardless.

So yeah, given this weird behaviour of mypy, it's better to keep `warn_on_return=true`